### PR TITLE
Make more use of expression functions

### PR DESCRIPTION
**Pull request by sschuberth**
For simplicity.

Signed-off-by: Sebastian Schuberth <sschuberth@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/1547)

### CODE REVIEW

**Comment by fviernau** [comment]
I'm aware that the style we currently use allows implicit return types of functions.
Anyhow, besides saving a few chars I don't see why implicit return types make things simpler. Personally I much prefer explicit over implicit return types for the below reasons:

- often times when reading a function I'm in the first place interrested in param and return type and implicit return time make this information harder to extract. Sometimes I'm even just interrested in the signature but not in the implementation
- changing the return type becomes less obvious. The IDE in some cases gives hints but in code reviews such change can easily go unnoticed. 
- adding some println or some line of code for debugging is easier
- see also the section about return types here: https://proandroiddev.com/why-kotlin-sucks-ab27a6b15cb6

I'd tend to do the opposite of this change and instead use explicit return types more often.
@mnonnenmacher , @sschuberth  what do you think?

**Comment by sschuberth** [comment]
Just to be sure @fviernau, you're not against using [expression functions](https://kotlinlang.org/docs/reference/functions.html#single-expression-functions) in general, but when doing so, you suggest to keep the optional return type explicitly, correct?

**Comment by fviernau** [comment]
> Just to be sure @fviernau, you're not against using [expression functions](https://kotlinlang.org/docs/reference/functions.html#single-expression-functions) in general, but when doing so, you suggest to keep the optional return type explicitly, correct?

yes, that's exactly what I meant.

**Comment by sschuberth** [comment]
Ok, I'd be fine with that and will change this PR accordingly. But what about expression function that return nothing? Would you prefer to omit the return type for those, or explicitly return `Unit`?

**Comment by sschuberth** [comment]
I have to retract my initial agreement to @fviernau's objections a bit: For consistency, I started to write out the return types for all existing expressions functions explicitly. In several cases this leads to code like

    fun collectConcludedLicenses(omitExcluded: Boolean = false): SortedMap<Identifier, SpdxExpression?> =
        sortedMapOf<Identifier, SpdxExpression?>().also { licenses ->

or

    fun toPackage(): Package = Package(...)

which is extremely verbose, and sometimes even requires the addition of imports. As I believe we'd need a clear rule, i.e. either always or never omit optional return types, I tend to say we should stick to the current rule and always omit optional return types.

@mnonnenmacher, what's your opinion?

**Comment by mnonnenmacher** [comment]
What about these rules:
* Do not use return types in expression functions, unless:
  * The return type should be a more generic class than what the expression returns, e.g. if the return type should be `SortedSet` but the expression returns `TreeSet`
  * The return type is not obvious from the expression, otherwise it makes code review of expression functions unnecessarily hard plus the other risks mentioned in the article @fviernau linked.
* Expression functions should be short and simple. E.g. a one-liner, or a single constructor or function call with one argument per line which is still simple, or a concatenation of functions like `filter {}.map {}...`, if their lambda arguments are simple.
* Do not "trick" complex functions into single expression functions using `also {}` or similar functions.

For example:
* The `toPackage()` function can be an expression function and does not need a return type, because the function is simple (even though it is more than one line) and the return type is obvious.
* For `collectConcludedLicenses(...)` I would not use an expression function anymore, even though the return type is obvious, but using `also` to "trick" the complex function into an expression function seems like a misuse of the feature to me now.
* For a function like the one below the return type should be defined (it currently isn't), because it is not obvious from the function itself, and changing the return type of `getAllDetectedLicenses()` would also change the return type of this function, which is risky.
```kotlin
fun getDetectedLicensesForId(id: Identifier): SortedSet<String> =
        scanner?.results?.scanResults?.find { it.id == id }.getAllDetectedLicenses()
```

**Comment by mnonnenmacher** [comment]
In general I think readability is more important than compactness. One more example for the first function in this PR. Instead of:

```kotlin
private fun extractDeclaredLicenses(node: JsonNode) =
    sortedSetOf<String>().apply {
        val license = node["pkgMeta"]["license"].textValueOrEmpty()
        if (license.isNotEmpty()) {
            add(license)
        }
    }
```

I would prefer:

```kotlin
private fun extractDeclaredLicenses(node: JsonNode): SortedSet<String> {
    val license = node["pkgMeta"]["license"].textValueOrEmpty()
    return if (license.isNotEmpty()) {
        sortedSetOf(license)
    } else {
        sortedSetOf()
    }
}
```

or:

```kotlin
private fun extractDeclaredLicenses(node: JsonNode): SortedSet<String> {
    val declaredLicenses = sortedSetOf<String>()
    val license = node["pkgMeta"]["license"].textValueOrEmpty()
    if (license.isNotEmpty()) {
        declaredLicenses.add(license)
    }
    return declaredLicenses
}
```

Because I find this easier to read (less nesting). Also, while it is technically a single expression, it is in fact a list of expressions hidden inside the lambda argument. And finally, the other variants are not longer than the single expression function.

**Comment by sschuberth** [comment]
I basically like these rules, expect for the fact that some of these leave room for interpretation: What "is not obvious from the expression" is quite subjective, and while "readability is more important than compactness" is certainly true, "readability" depends on your programming background (your examples start to look more like Java than like Kotlin, so they're certainly more readable for someone with a Java background).

Whatever conventions we come up with, we should probably add them to `CONTRIBUTING.md`.

In the meantime, I'll try to make this PR adhere to @mnonnenmacher's rules.

**Comment by fviernau** [comment]
I'm still worried about the following case. If someone makes a commit which changes the return type of someOtherFun, from that commit it will not be obvious at all that also the return type of someFun is changed. Thus it can easily go unnoticed in the review, which I personally don't like.
But if we say in this case the return type is not obvious and thus it needs to be explicitly written down
then I'm like the rules as well. Agree with Sebastian that room for intepretation should be minimized.
`
fun someFun(a: Int) = someOtherFun(a + 1)
`
Would you say that the return type obvious here or not?

**Comment by mnonnenmacher** [comment]
Yes, the wording of my proposal is not yet good enough, let's discuss the details in the PR for `CONTRIBUTING.md`. @sschuberth Do you want to create that, or shall I?

@fviernau : In your example I would say yes, it needs a return type, because you need to lookup `someOtherFun` to know what the return type is, that's what I meant with "not obvious", in constrast to e.g. `fun someFun(a: Int) = a + 1`.

@sschuberth : The examples still look like Kotlin, because they use `val` and no semicolons :) I think it is hard to find the right balance between functional and imperative style and agree readability depends on experience with the language, but I find long code blocks and deep nesting always harder to understand than short plain code blocks.

In specific, comparing these two variants I find the second definitely easier to read, even after two years with Kotlin:

```kotlin
sortedMapOf<Identifier, SpdxExpression?>().also { licenses ->
   licenses.add(license)
}
```

```kotlin
val licenses = sortedMapOf<Identifier, SpdxExpression?>()
licenses.add(license)
```

I have used the first variant in many places in the code because I wanted to use expression functions and use lambdas, but I have changed my mind about that in the meantime.

Anyway, we should not overdo it with the rules and more treat them as guidelines, because it's hard to come up with rules that match any possible case, so some ambiguities should be fine.

Btw, this is the function in Java, even without functional idioms the Kotlin version looks better :)

```java
    private SortedSet<String> extractDeclaredLicenses(JsonNode node) {
        final SortedSet<String> declaredLicenses = new TreeSet<>();
        final String license = textValueOrEmpty(node.get("pkgMeta").get("license"));
        if (!license.isEmpty()) {
            declaredLicenses.add(license);
        }
        return declaredLicenses;
    }
```

**Comment by sschuberth** [comment]
> If someone makes a commit which changes the return type of someOtherFun, from that commit it will not be obvious at all that also the return type of someFun is changed. Thus it can easily go unnoticed in the review, which I personally don't like.

@fviernau, I'd argue this doesn't really go unnoticed as also all callers of `someFun` (that use it's return type) would have to be adjusted.

> Do you want to create that, or shall I?

You please, @mnonnenmacher  😀 

> Anyway, we should not overdo it with the rules and more treat them as guidelines, because it's hard to come up with rules that match any possible case, so some ambiguities should be fine.

Agreed!

**Comment by fviernau** [comment]
>  this doesn't really go unnoticed as also all callers of `someFun` (that use it's return type) would have to be adjusted.

You're right, that's actually an interresting point. So only in case there is no caller at all or all callers are not in the ort code base this would go unnoticed. 
So in short: how about relaxing the rule like said already

> Anyway, we should not overdo it with the rules and more treat them as guidelines, because it's hard to come up with rules that match any possible case, so some ambiguities should be fine.

but require explicit return types always for classes we consider as public API - strong candidate are classes interfacing with the evaluator kotlin script / notice postprocessing script.
Maybe as a side benefit we would move towards more clarity what exactly is part of our public API.

**Comment by sschuberth** [comment]
> Maybe as a side benefit we would move towards more clarity what exactly is part of our public API.

True. So far our "silent agreement" was that only the public functions of the respective module's main class (*analyzer* module -> `Analyzer.kt`, *downloader* module -> `Download.kt`, etc.) are our public API.

But until we've clarified and documented that, it this isolated PR now good enough for you, @fviernau?

File: analyzer/src/main/kotlin/managers/PhpComposer.kt [hunk 1]
```diff
@@ -312,18 +312,16 @@ class PhpComposer(
     private fun parseDeclaredLicenses(packageInfo: JsonNode) =
         packageInfo["license"]?.mapNotNull { it?.textValue() }?.toSortedSet() ?: sortedSetOf<String>()
 
-    private fun parseVcsInfo(packageInfo: JsonNode): VcsInfo {
-        return packageInfo["source"]?.let {
+    private fun parseVcsInfo(packageInfo: JsonNode) =
```
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/PhpComposer.kt]: prefer explicit return type here

File: analyzer/src/main/kotlin/managers/PhpComposer.kt [hunk 2]
```diff
@@ -312,18 +312,16 @@ class PhpComposer(
...
+        packageInfo["source"]?.let {
             VcsInfo(it["type"].textValueOrEmpty(), it["url"].textValueOrEmpty(), it["reference"].textValueOrEmpty())
         } ?: VcsInfo.EMPTY
-    }
 
-    private fun parseArtifact(packageInfo: JsonNode): RemoteArtifact {
-        return packageInfo["dist"]?.let {
+    private fun parseArtifact(packageInfo: JsonNode) =
```
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/PhpComposer.kt]: prefer explicit return type here

File: analyzer/src/main/kotlin/managers/Pip.kt [hunk 3]
```diff
@@ -112,8 +112,8 @@ object PythonVersion : CommandLineTool {
      * different Python versions can by installed in arbitrary locations, and the Python executable is even usually
      * called the same in those locations.
      */
-    fun getPythonInterpreter(version: Int): String {
-        return if (Os.isWindows) {
+    fun getPythonInterpreter(version: Int) =
```
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/Pip.kt]: prefer explicit return type here

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt [hunk 4]
```diff
@@ -50,10 +50,9 @@ object AnalyzerCommand : CommandWithHelp() {
             val PACKAGE_MANAGERS = PackageManager.ALL.associateBy { it.managerName.toUpperCase() }
         }
 
-        override fun convert(name: String): PackageManagerFactory {
-            return PACKAGE_MANAGERS[name.toUpperCase()]
+        override fun convert(name: String) =
```
Comment (fviernau) [inline cli/src/main/kotlin/commands/AnalyzerCommand.kt]: prefer explicit return type here

File: cli/src/main/kotlin/commands/ReporterCommand.kt [hunk 5]
```diff
@@ -54,10 +54,9 @@ object ReporterCommand : CommandWithHelp() {
             val REPORTERS = Reporter.ALL.associateBy { it.reporterName.toUpperCase() }
         }
 
-        override fun convert(name: String): Reporter {
-            return REPORTERS[name.toUpperCase()]
+        override fun convert(name: String) =
```
Comment (fviernau) [inline cli/src/main/kotlin/commands/ReporterCommand.kt]: prefer explicit return type here

File: downloader/src/main/kotlin/vcs/Cvs.kt [hunk 6]
```diff
@@ -150,17 +150,15 @@ class Cvs : VersionControlSystem(), CommandLineTool {
                         decimals.dropLast(1).last() == "0"
             }
 
-            override fun listRemoteBranches(): List<String> {
-                return listSymbolicNames().mapNotNull { (name, version) ->
+            override fun listRemoteBranches() =
```
Comment (fviernau) [inline downloader/src/main/kotlin/vcs/Cvs.kt]: prefer explicit return type here

File: downloader/src/main/kotlin/vcs/Cvs.kt [hunk 7]
```diff
@@ -150,17 +150,15 @@ class Cvs : VersionControlSystem(), CommandLineTool {
...
+                listSymbolicNames().mapNotNull { (name, version) ->
                     if (isBranchVersion(version)) name else null
                 }
-            }
 
-            override fun listRemoteTags(): List<String> {
-                return listSymbolicNames().mapNotNull { (name, version) ->
+            override fun listRemoteTags() =
```
Comment (fviernau) [inline downloader/src/main/kotlin/vcs/Cvs.kt]: prefer explicit return type here

File: scanner/src/main/kotlin/scanners/BoyterLc.kt [hunk 8]
```diff
@@ -154,13 +154,12 @@ class BoyterLc(name: String, config: ScannerConfiguration) : LocalScanner(name,
         }
     }
 
-    override fun getRawResult(resultsFile: File): JsonNode {
-        return if (resultsFile.isFile && resultsFile.length() > 0L) {
+    override fun getRawResult(resultsFile: File) =
```
Comment (fviernau) [inline scanner/src/main/kotlin/scanners/BoyterLc.kt]: prefer explicit return type here

File: scanner/src/main/kotlin/scanners/FileCounter.kt [hunk 9]
```diff
@@ -68,13 +68,12 @@ class FileCounter(name: String, config: ScannerConfiguration) : LocalScanner(nam
         return ScanResult(Provenance(), getDetails(), summary, result)
     }
 
-    override fun getRawResult(resultsFile: File): JsonNode {
-        return if (resultsFile.isFile && resultsFile.length() > 0L) {
+    override fun getRawResult(resultsFile: File) =
```
Comment (fviernau) [inline scanner/src/main/kotlin/scanners/FileCounter.kt]: prefer explicit return type here

File: scanner/src/main/kotlin/scanners/Licensee.kt [hunk 10]
```diff
@@ -126,13 +126,12 @@ class Licensee(name: String, config: ScannerConfiguration) : LocalScanner(name,
         }
     }
 
-    override fun getRawResult(resultsFile: File): JsonNode {
-        return if (resultsFile.isFile && resultsFile.length() > 0L) {
+    override fun getRawResult(resultsFile: File) =
```
Comment (fviernau) [inline scanner/src/main/kotlin/scanners/Licensee.kt]: prefer explicit return type here

File: scanner/src/main/kotlin/scanners/ScanCode.kt [hunk 11]
```diff
@@ -335,13 +335,12 @@ class ScanCode(name: String, config: ScannerConfiguration) : LocalScanner(name,
         }
     }
 
-    override fun getRawResult(resultsFile: File): JsonNode {
-        return if (resultsFile.isFile && resultsFile.length() > 0L) {
+    override fun getRawResult(resultsFile: File) =
```
Comment (fviernau) [inline scanner/src/main/kotlin/scanners/ScanCode.kt]: prefer explicit return type here

**Comment by sschuberth** [comment]
> I've commented on each function which did not match our preferences.

I'll comment on (some of) these why I chose to not have an explicit return type. This does not mean I'm refusing to add it, but I simply want to explain myself to provide another view, as the rules are not finalized yet.

File: analyzer/src/main/kotlin/managers/PhpComposer.kt (hunk 1 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/PhpComposer.kt]: For me, the return type is obvious here as the "else" branch has `VcsInfo.EMPTY`.

File: analyzer/src/main/kotlin/managers/PhpComposer.kt (hunk 2 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/PhpComposer.kt]: Similar to above.

File: analyzer/src/main/kotlin/managers/Pip.kt (hunk 3 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Pip.kt]: Done.

File: cli/src/main/kotlin/commands/ReporterCommand.kt (hunk 5 above)
Comment (sschuberth) [reply cli/src/main/kotlin/commands/ReporterCommand.kt]: That's probably an edge case as the lookup happens in a map called `REPORTERS`, so it's quite obvious, but I've added the `Reporter` return type nonetheless.

File: downloader/src/main/kotlin/vcs/Cvs.kt (hunk 6 above)
Comment (sschuberth) [reply downloader/src/main/kotlin/vcs/Cvs.kt]: This clearly returns a list of names, and names are usually strings, so to me the return type is obvious.

File: downloader/src/main/kotlin/vcs/Cvs.kt (hunk 7 above)
Comment (sschuberth) [reply downloader/src/main/kotlin/vcs/Cvs.kt]: Similar to above.

File: scanner/src/main/kotlin/scanners/BoyterLc.kt (hunk 8 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/BoyterLc.kt]: For me, the return type is obvious here as the "else" branch has `EMPTY_JSON_NODE`.

File: scanner/src/main/kotlin/scanners/FileCounter.kt (hunk 9 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/FileCounter.kt]: Similar to above.

File: scanner/src/main/kotlin/scanners/Licensee.kt (hunk 10 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/Licensee.kt]: Similar to above.

File: scanner/src/main/kotlin/scanners/ScanCode.kt (hunk 11 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/ScanCode.kt]: Similar to above.

**Comment by mnonnenmacher** [comment]
> > I've commented on each function which did not match our preferences.
> 
> I'll comment on (some of) these why I chose to not have an explicit return type. This does not mean I'm refusing to add it, but I simply want to explain myself to provide another view, as the rules are not finalized yet.

Our approach was to check how long it takes us to find out what the return type is, and if this is significantly longer than it would take to write down the return type (usually ~1s), because usually you read the code far more often than you change it. However, most times I read the sources in the IDE where return types are augmented, so it's mainly relevant for code review.
What we need to decide is, if it takes me 5-10s to figure out the return type, wouldn't it be better to write it down? Besides that public API should always have the types defined to avoid accidental API changes.
The other approach was to check if you need context knowledge to know the return type. For example, you need to know that `EMPTY_JSON_NODE` is a `JsonNode`, or that Jackson's `readTree` returns a `JsonNode`, otherwise you need to investigate first.

In the end I don't have a strong opinion here, I would also be ok with merging the PR as is.

**Comment by sschuberth** [comment]
> In the end I don't have a strong opinion here, I would also be ok with merging the PR as is.

I'd be glad if you'd approve to merge this in the current state, simply because I'm too "lazy" to rework this PR once more, esp. as I currently would like to spend my time on more important changes.

We'd need to review the current code base according to the rules we come up with anyway, and any necessary changes in the code in this PR could be done in that context.

**Comment by sschuberth** [comment]
Thanks @fviernau!

**Pull request merged**